File: test-suite/handwritten-src/ts/CppExceptionTest.ts
import {TestCase, allTests, assertEq, assertTrue} from "./testutils"
import * as test from "../../generated-src/ts/test";
import {DjinniModule} from "@djinni_support/DjinniModule"

class TsThrower implements test.ThrowingInterface {
    throwException() {
        throw new Error("TS Error");
    }
}

class CppThrower implements test.ThrowingInterface {
    cpp: test.CppException;
    constructor(cpp: test.CppException) {
        this.cpp = cpp;
    }
    throwException() {
        this.cpp.throwAnException();
    }
}

class CppExceptionTest extends TestCase  {
    m: test.Test_statics & DjinniModule;
    cppInterface: test.CppException;

    constructor(module: test.Test_statics & DjinniModule) {
        super(module);
        this.m = module;
        this.cppInterface = this.m.testsuite.CppException.get()!;
    }

    // C++ exception can be handled in TS
    testCppException() {
        var thrown = null;
        try {
            this.cppInterface.throwAnException();
        } catch (e: any) {
            thrown = e;
        }
        assertTrue(thrown instanceof Error);
        assertEq(thrown.message, "C++: Exception Thrown");
    }

    // TS exception passes through C++ code and handled in TS higher up in the stack
    testTsExceptionPassthrough() {
        var thrown = null;
        try {
            this.cppInterface.callThrowingInterface(new TsThrower());
        } catch (e: any) {
            thrown = e;
        }
        assertTrue(thrown instanceof Error);
        assertEq(thrown.message, "TS Error");
    }

    // TS exception can be handled in C++
    testTsExceptionHandledByCpp() {
        let res = this.cppInterface.callThrowingAndCatch(new TsThrower());
        // the stack info from the js exception
        assertTrue(res.length > 0 );
    }

    // C++ exception passes through TS code and handled in C++ higher up in the stack
    testCppExceptionPassthrough() {
        let res = this.cppInterface.callThrowingAndCatch(new CppThrower(this.cppInterface));
        assertEq(res, "Exception Thrown");
    }
}

allTests.push(CppExceptionTest);

File: test-suite/handwritten-src/swift/CppExceptionTest.swift
import XCTest
import DjinniSupport
@testable import TestSuite

final class CppExceptionTest: XCTestCase {

    class SwiftThrower: ThrowingInterface {
        func throwException() throws -> Void {
            throw DjinniError("Swift Error")
        }
    }

    let cppInterface = (try? CppException_statics.get())!
    
    func testCppException() throws {
        var thrown: DjinniError? = nil
        do {
            let _ = try cppInterface.throwAnException()
        } catch let e as DjinniError {
            thrown = e
        }
        XCTAssertEqual(try XCTUnwrap(thrown).errorMessage, "Exception Thrown")
    }

    func testSwiftExceptionHandledByCpp() throws {
        let res = try cppInterface.callThrowingAndCatch(SwiftThrower())
        XCTAssertEqual(res, "Swift Error")
    }
}
